src/Workleap.DomainEventPropagation.Subscription.PullDelivery.Tests/EventPullerTests.cs

```diff
@@ -1,58 +1,258 @@
-﻿using Azure.Messaging.EventGrid.Namespaces;
+﻿using AutoBogus;
+using Azure.Messaging;
+using Azure.Messaging.EventGrid.Namespaces;
 using FakeItEasy;
-using Microsoft.Extensions.Azure;
+using FakeItEasy.Configuration;
+using Microsoft.Extensions.Hosting;
 using Microsoft.Extensions.Logging.Abstractions;
 using Microsoft.Extensions.Options;
+using Workleap.DomainEventPropagation.EventGridClientAdapter;
 
 namespace Workleap.DomainEventPropagation.Subscription.PullDelivery.Tests;
 
+// Test class cannot be made static
+#pragma warning disable CA1052
 public class EventPullerTests
+#pragma warning restore CA1052
 {
-    [Fact]
-    public async Task GivenPullerWithTwoSubscribers_AfterStarting_EveryRegisteredClientWasCalled()
+    private static IEventGridClientAdapter GivenFakeClient(IEventGridClientWrapperFactory clientWrapperFactory, string subName)
     {
-        // Given
-        const string sub1 = "Subscriber#1";
-        const string sub2 = "Subscriber#2";
-        var eventGridClientDescriptors = new EventGridClientDescriptor[] { new(sub1), new(sub2) };
-
-        var clientFactory = A.Fake<IAzureClientFactory<EventGridClient>>();
-        var fakeClient1 = A.Fake<EventGridClient>();
-        var fakeClient2 = A.Fake<EventGridClient>();
-        A.CallTo(() => clientFactory.CreateClient(sub1)).Returns(fakeClient1);
-        A.CallTo(() => clientFactory.CreateClient(sub2)).Returns(fakeClient2);
-
-        var optionMonitor = A.Fake<IOptionsMonitor<EventPropagationSubscriptionOptions>>();
-        var options1 = new EventPropagationSubscriptionOptions() { TopicName = "topic1", SubscriptionName = "sub1" };
-        var options2 = new EventPropagationSubscriptionOptions() { TopicName = "topic2", SubscriptionName = "sub2" };
-        A.CallTo(() => optionMonitor.Get(sub1)).Returns(options1);
-        A.CallTo(() => optionMonitor.Get(sub2)).Returns(options2);
-
-        // When
-        var puller = new EventPuller(eventGridClientDescriptors, clientFactory, optionMonitor, new NullLogger<EventPuller>());
+        var fakeClient = A.Fake<IEventGridClientAdapter>();
+        A.CallTo(() => clientWrapperFactory.CreateClient(subName)).Returns(fakeClient);
+        return fakeClient;
+    }
+
+    private static EventPropagationSubscriptionOptions GivenOption(IOptionsMonitor<EventPropagationSubscriptionOptions> optionMonitor, string clientName)
```

Comment:
`Option` is too generic, I would suggest to use the full name to make it explicit:

```suggestion
    private static EventPropagationSubscriptionOptions GivenEventPropagationSubscriptionOptions (IOptionsMonitor<EventPropagationSubscriptionOptions> optionMonitor, string clientName)
```